### PR TITLE
execution/state: memoize the access-list slot set per address

### PR DESCRIPTION
**Pull request by AskAlexSharov**
Every `SLOAD`/`SSTORE` resolves the address's slot set through the `addresses` map, then probes the slot map. A call frame only ever touches its own storage — consecutive storage ops resolve the same set, and hot loops re-touch the same slot. So `accessList` memoizes the last resolved `(address -> slot set)` and the last warm slot in it: the same address skips the address lookup, and the same `(address, slot)` pair is answered by two interned-handle compares, zero map probes. `IntraBlockState.SlotKnownWarm` exposes the pair-compare to the SLOAD/SSTORE gas functions, cheap enough to inline.

### Numbers

`BenchmarkAccessListSlots`, n0, 10 interleaved rounds, `benchstat` `main` vs this branch. `run64` = 64 consecutive storage ops on one address (a long call frame); `run1` = the address changes on every op, so the memo never hits and only pays its compare:

| | main | this PR | Δ |
|---|---|---|---|
| `warm-hit/run64` | 8.26ns | 5.99ns | **-27.5%** |
| `warm-hit-hot/run64` | 8.56ns | 2.16ns | **-74.7%** |
| `contains/run64` | 7.36ns | 6.74ns | -8.5% |
| `cold-insert/run64` | 16.34ns | 14.97ns | -8.4% |
| `warm-hit/run1` | 8.31ns | 9.30ns | **+11.9%** |
| `warm-hit-hot/run1` | 8.54ns | 9.59ns | **+12.2%** |
| `cold-insert/run1` | 16.26ns | 17.24ns | **+6.0%** |
| geomean, all 16 arms | 9.58ns | 7.25ns | **-24.3%** |

The `run1` regression is the memo's worst case; producing it in real execution takes a `CALL` between every two storage ops, and the frame setup dwarfs the ~1ns.

<details>
<summary>Details: promotion, invalidation, correctness</summary>

- A slot found warm by a map probe is *promoted* into the memo, so repeat-read loops converge onto the two-compare path.
- `SlotKnownWarm` is conservative: `true` means warm, `false` means unknown — callers fall back to `AddSlotToAccessList`, which records warmth and charges the cold cost.
- The memo is dropped in `Reset` and `DeleteSlot` — the only paths that can rebind a slot map away from its address. `DeleteSlot` must also clear the warm-slot memo, or a journal revert could leave the reverted slot looking warm; `TestAccessListWarmSlotMemoDroppedOnDelete` pins that.
</details>

<details>
<summary>End-to-end numbers</summary>

`execution/vm/benchmark`, same rounds, vs `main`: suite geomean -0.8%. Storage-access arms -2…-10%: `SLOADWarm` -5.2…-8.0%, `StorageKeyReuse` -6.4…-9.5%, `ERC20BatchTransfers` -3.3…-6.2%, `ERC20BalanceOf` -4.3%, `SLOADCold` -1.9…-5.2%, `SSTORE` -1.8…-2.8%.

`PureArithmetic` shows +10%, but it executes none of the touched code and the same arms return to ~0% vs main at the next branch in the stack (#22744), which only shifts code placement further — layout noise, not a cost of this change.
</details>

<details>
<summary>Rejected alternatives</summary>

- One flat map keyed by `struct{addr, slot}` (would delete the index/`-1` sentinel/LIFO-truncation machinery): **87% slower** (16.10ns vs 8.62ns geomean). Two 8-byte interned-handle probes hit Go's `map_fast64` specialization; one 16-byte pointer-containing struct key falls back to generic typehash.
- Deriving `slotChange` from a `len(map)` delta to insert with a single `mapassign`: a warm hit through `mapassign` costs ~20% more than through `mapaccess`, and warm hits dominate — probe-then-insert stays.
</details>

All PR checks are green on the current head, including the EEST/consensus spec suites (the gas-metering gate), race tests, hive, and lint.

### CODE REVIEW

**Comment by AskAlexSharov** [comment]
Addressed the comment on the struct doc — reworded to say `nil lastSlots` means *no memo* and that address-only entries never populate it, which is what the `slotmap != nil` guard actually implements.

**Comment by AskAlexSharov** [comment]
Corrected the numbers. The `cold-insert` arm in the first version of the benchmark covered only 64 of 256 (address, slot) pairs for `runLen` 1/4/16 — three quarters of those timed iterations were warm re-hits, so that row understated cold-insert cost by ~40%. Fixed formula walks all 256 pairs per Reset window while preserving the intended run length; base cold-insert is now ~16.4ns rather than ~10ns, and `cold-insert/run1` is neutral rather than the +3.72% originally reported.

Also made `Contains` read the memo without writing it, which removes its `run1` regression (+14.41% → -6.08%) and keeps the query path non-mutating. Re-measured on a 16-core EPYC with `-count=10`; body updated.

File: execution/state/access_list.go [hunk 1]
```diff
@@ -32,6 +32,13 @@ import (
 type accessList struct {
 	addresses map[accounts.Address]int
 	slots     []map[accounts.StorageKey]struct{}
+
+	// Memo of the last resolved (address -> slot set); nil lastSlots means empty.
+	// A call frame runs every SLOAD/SSTORE against one address, so consecutive
+	// storage ops reuse it instead of re-reading addresses. Must be dropped
+	// whenever a slot map can be handed to a different address.
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/state/access_list.go]: The comment already reads "a nil `lastSlots` means no memo" — not "means empty" — so it matches the `slotmap != nil` fast path exactly. Address-only entries (`idx == -1`) fall through to the slow path and set the memo there, which the comment does not contradict. No change needed.

File: execution/state/access_list.go [hunk 2]
```diff
@@ -147,6 +180,8 @@ func (al *accessList) DeleteSlot(address accounts.Address, slot accounts.Storage
 		}
 		al.slots = al.slots[:idx]
 		al.addresses[address] = -1
+		// The truncated map is now free for a different address to claim.
+		al.dropMemo()
 	}
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/state/access_list.go:194]: Added `TestAccessListMemoDroppedOnSlotMapReuse`.

Worth recording what it took to make it actually guard the invariant: `Contains` reads through the memo too, so the straightforward version (add → delete → re-add → assert) passes even with `dropMemo()` deleted — the write and the read both go through the same stale map and agree with each other. The test only discriminates once a second address displaces the memo and forces the reads onto the slow path. Verified by deleting the `dropMemo()` call: red with it gone ("addrA's slot was written into a detached slot map"), green with it back.

**Comment by AskAlexSharov** [comment]
## End-to-end EVM benchmarks

The numbers in the description measure `accessList` in isolation. Here is what the change does to actual EVM execution, including Snailtracer.

**n0** (16-core EPYC 4344P, idle), base = merge-base `bafdb4d5cf`, head = `0d14f001c9`.

A first sequential run (all of base, then all of head) reported `execution/vm/runtime` at **+2.36% geomean**, with `SimpleLoop/loop2` at +36%. That was machine drift, not this change — variance on the second half was ±16–26% against ±0–2% on the first, and the box's load rose from 0.00 to 2.1 between the halves. Everything below is therefore **interleaved**: both test binaries built once with `go test -c`, then run round-robin `base, new, base, new…` for 8 rounds, so drift hits both refs equally. Under that method the `runtime` "regression" disappears (`loop2` +36.16% → `~`, geomean **−0.25%**, every entry `~`).

### State access — the path this PR touches

```
SLOADWarm/10slots                -10.30% (p=0.003)    SLOADCold/500slots     -7.42% (p=0.000)
SLOADWarm/50slots                 -8.96% (p=0.000)    SLOADCold/100slots     -5.76% (p=0.000)
SLOADWarm/500slots                -7.62% (p=0.000)    SLOADCold/50slots      -5.14% (p=0.001)
SLOADWarm/100slots                -6.15% (p=0.003)    SLOADCold/10slots      -4.16% (p=0.000)

StorageDiversity/100slots         -7.67% (p=0.001)    SSTORE/nonzero-to-nonzero  -4.12% (p=0.001)
StorageDiversity/1000slots        -7.07% (p=0.001)    SSTORE/zero-to-nonzero     -3.75% (p=0.000)
ERC20BalanceOf/balanceOf/100M     -7.19% (p=0.000)    SSTORE/nonzero-to-zero     -2.86% (p=0.010)
ERC20BatchTransfers/batch-10      -4.44% (p=0.000)    ERC20Transfer/100M         -3.95% (p=0.021)
ERC20BatchTransfers/batch-50      -3.90% (p=0.010)    ERC20TransferFrom/100M     -3.94% (p=0.000)
ERC20BatchTransfers/batch-5       -2.80% (p=0.001)
```

geomean of that group **−3.45%**.

`TransientStorage/*` is `~` on all three arms, which is the control you want: `TLOAD`/`TSTORE` never enter the access list, so no effect there is the correct result.

### Cost: cross-contract call chains

```
NestedStaticCalls/depth-4         +3.30% (p=0.002)
CallWithValue/no-value            +3.20% (p=0.015)
DelegateCallProxy/4-layers        +2.92% (p=0.002)
NestedStaticCalls/depth-8         +1.92% (p=0.001)
```

This is the `run1` case from the description, now visible end to end: a call chain changes address on every frame, so the memo never hits and pays the compare. `depth-2`, `depth-16`, `DelegateCallProxy/1-2 layers` and `DeFiSwapChain` are all `~`.

### Snailtracer, and a caveat on everything under ~3%

```
Snailtracer     -1.29% (p=0.015)   interleaved, -benchtime=1s x8
Snailtracer     -2.11% (p=0.000)   dedicated run, -benchtime=10s x10
```

Directionally consistent, and it matches the profile — `AddSlotToAccessList` is ~2% of Snailtracer's runtime, and Snailtracer is a single contract touching its own storage, so the memo hits ~always.

That said: **code placement alone moves `execution/vm/benchmark` by roughly ±3% on these boxes**, and `Snailtracer`, `MixedCompute` and `Keccak256` are among the layout-sensitive ones. So the Snailtracer figure, the call-chain regressions above, and these

```
Keccak256/32B/100M   +1.46% (p=0.007)     MixedCompute/100M   -2.74% (p=0.000)
Keccak256/256B/100M  +1.55% (p=0.000)     Keccak256/4KB/100M  -1.21% (p=0.000)
```

should be read as "inside the layout band". `Keccak256` cannot touch the access list at all, so its ±1.5% is placement, most likely from `accessList` gaining two fields and shifting the offsets of everything after it in `IntraBlockState`. Anyone wanting to settle a sub-3% item should use a relocation control (a `//go:noinline` pad above the function under test, verified instruction-identical with `objdump`) rather than a plain A/B.

The claim this PR should rest on is the state-access group: −3% to −10%, well outside that band, with a coherent gradient — the saving tracks memo hit rate, largest where one address is hit repeatedly (`SLOADWarm/10slots`, `ERC20BalanceOf`) and smaller as addresses churn.

**Review by sudeepdino008 — APPROVED** [review]

**Review by yperbasis — DISMISSED** [review]
Requesting changes for one correctness issue in the memo fast path.

lastAddr defaults and resets to accounts.NilAddress, but Contains, AddSlot, and SlotKnownWarm treat address equality alone as proof that a memo exists. On a new or reset access list, Contains(accounts.NilAddress, accounts.ZeroKey) reports the address as present, while AddSlot(accounts.NilAddress, accounts.ZeroKey) panics with assignment to entry in nil map. SlotKnownWarm(accounts.NilAddress, accounts.NilKey) also reports a pair as warm even though it was never added.

Please gate all three fast paths on lastSlots != nil, or add an explicit memo-valid flag, and add a regression test for the empty/reset state.

**Comment by AskAlexSharov** [comment]
Fixed in f162ac76eb.

All three fast paths (`Contains`, `AddSlot`, `SlotKnownWarm`) now require `lastSlots != nil`, so a list with no memo always takes the slow path. `lastSlots` is the gate rather than a separate flag because it is the map `AddSlot` writes into — the nil check is exactly the condition that made that write unsafe, and `SlotKnownWarm` still inlines with it.

Regression tests in `access_list_test.go`, both red before the fix:

- `TestAccessListMemoOnEmptyList` — fresh and post-`Reset` lists: `Contains(NilAddress, ZeroKey)`, `Contains(NilAddress, NilKey)`, and `AddSlot(NilAddress, ZeroKey)`, which panicked.
- `TestSlotKnownWarmOnEmptyAccessList` — the same invariant at the gas-charge entry point.

Main is merged in dc32709d8d. The conflict was in `makeGasSStoreFunc`: main now charges the access cost before adding the slot to the access list, so the merged version keeps that order and uses the memo only as a fast path — `SlotKnownWarm`, then the authoritative `SlotInAccessList` probe on a memo miss, then `AddSlotToAccessList` after the affordability check.

**Comment by yperbasis** [comment]
Remaining items, all non-blocking except the rebase:

1. Rebase needed: `execution/vm/operations_acl.go` conflicts with #22748 (`peekStorageKey` now takes `evm`).

2. Warm `Contains` hits never promote the memo (Copilot's suppressed comment points at the same thing). The SSTORE path (`SlotKnownWarm` miss → `SlotInAccessList`) neither promotes nor rebinds, so warm SSTOREs without a preceding same-slot SLOAD — blind writes, slots pre-warmed via a tx access list, loops after a displacing call — keep paying map probes instead of converging on `SlotKnownWarm`. No regression vs main, since read-modify-write still converges via the SLOAD's `AddSlot`. A minimal fix that keeps `Contains` non-mutating: in `makeGasSStoreFunc`, key the post-OOG add off the `SlotKnownWarm` result instead of `slotPresent`, so `AddSlotToAccessList` runs as a promoting no-op on warm-via-map hits; ideally with a benchmark arm for that two-call sequence. Alternatively, just scope the PR-body convergence claim to the SLOAD path.

3. Test pin gap: the reset arm of `TestAccessListMemoOnEmptyList` only queries `NilAddress`, so it does not pin the `dropMemo()` in `Reset` — removing that call is caught only by `TestRecreateAndRewind`, as a gas-used-vs-header panic far from the root cause (verified locally). Suggestion: after `AddSlot(addrA, slot1); Reset()`, also expect `Contains(addrA, slot1) == (false, false)` and `AddSlot(addrA, slot1) == (true, true)`.
